File: python/140. Word Break II.py
from collections import defaultdict
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: Set[str]
        :rtype: List[str]
        """
        dic = collections.defaultdict(list)
        
        def helper(s):
            if not s:
                return [None]
            if s in dic:
                return dic[s]
            res = []
            for word in wordDict:
                n = len(word)
                if word == s[:n]:
                    for each in helper(s[n:]):
                        if each:
                            res.append(word+" "+each)
                        else: 
                            res.append(word)
                dic[s] = res
            return res
        return helper(s)
            
            

File: cpp/101-150/140. Word Break II.cc
class Solution {
private:
    unordered_map<string, vector<string>> m;
    unordered_map<string, bool> dict;
    vecotr<string> combine(string word, vector<string> &prev) {
        for (int i = 0 ; i < prev.size(); i++) {
            prev[i] += " " + word;
        }
        return prev;
    }
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
    	for(string word : wordDict){
    		dict[word] = true;
    	}
    	return helper(s, dict);
    }
    vector<string> helper(string s, unordered_map<string, bool> &dict){
    	if(m[s] != m.end())
    		return m[s];
    	vector<string> res;
    	if(dict[s])
    		res.push_back(s);
    	for(int i = 1; i < s.size(); i++){
    		string word = s.substr(i);
    		if(dict[word] != dict.end()){
    			string rem = s.substr(0, i);
    			vector<string> prev = combine(word, helper(rem, dict));
    			res.insert(res.end(), prev.begin(), prev.end());
    		}
    	}
    	m[s] = res;
    	return res;
    }
};